### PR TITLE
Node

### PR DESCRIPTION
**Pull request by jbianquetti-nami**
Deployment that boots an NodeJS container, clones an git repo with a defined git branch and boots the NodeJS application. 

IMHO, this is a showroom Chart because it seems to me too simple.

Anyway, right now  it can be used as a demo chart.

### CODE REVIEW

File: incubator/node/templates/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    metadata:
+      labels:
+        app: {{ template "fullname" . }}
+    spec:
+      containers:
+      - name: {{ template "fullname" . }}
+        image: "{{ .Values.image }}:{{ .Values.imageTag }}"
+        imagePullPolicy: {{ default "" .Values.imagePullPolicy | quote }}
+        command: [ "sh" ]
+        args: [ "-c" , "git clone {{ .Values.gitRepository }} /app &&  cd /app &&  git checkout {{ .Values.gitBranch}}  &&  npm install &&  npm start" ]
```
Comment (prydonius) [inline incubator/node/templates/deployment.yaml]: It would be better to do the git clone in an init container, or using the git volume plugin http://kubernetes.io/docs/user-guide/volumes/#gitrepo

Comment (jbianquetti-nami) [inline incubator/node/templates/deployment.yaml]: Makes sense to me use the gitrepo volume plugin so there's not any need to attach a PV.
I'm gonna update the PR now.

File: incubator/node/README.md [hunk 2]
```diff
@@ -50,8 +50,8 @@ The following tables lists the configurable parameters of the node chart and the
 | `image`                              | node image                               | `bitnami/node:{VERSION}`                                |
 | `imageTag`                           | node image Tag                           | ``                                                      |
 | `imagePullPolicy`                    | Image pull policy                        | `Always` if `imageTag` is `latest`, else `IfNotPresent` |
-| `gitRepository`                      | Repo of the application                  | `git@github.com:jbianquetti-nami/simple-node-app.git`   |
-| `gitBranch`                          | Repo branch to checkout                  | `master`                                                |
+| `repository`                         | Repo of the application                  | `git@github.com:jbianquetti-nami/simple-node-app.git`   |
+| `revison`                            | Revision  to checkout                    | `master`                                                |
```
Comment (sameersbn) [inline incubator/node/README.md]: typo

File: incubator/node/Chart.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+name: node
+version: 0.1.0
+description: NodeJs  Event-driven I/O server-side JavaScript environment based on V8
```
Comment (prydonius) [inline incubator/node/Chart.yaml]: remove the `NodeJs `

File: incubator/node/Chart.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+- nodejs
+- git
+home: http://nodejs.org/
+sources:
+- https://github.com/bitnami/bitnami-docker-node
+maintainers:
+- name: Bitnami
+  email: containers@bitnami.com
+engine: gotpl
+icon:
```
Comment (prydonius) [inline incubator/node/Chart.yaml]: point to the bitnami node icon?

File: incubator/node/README.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
+# node
```
Comment (prydonius) [inline incubator/node/README.md]: capitalise

File: incubator/node/README.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+```console
+$ helm install .
+```
+
+## Introduction
+
+This chart bootstraps a [Node](https://github.com/bitnami/bitnami-docker-node) deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.
+
+It clones and deploy a nodejs application  from a git repository and  defined revision.
```
Comment (prydonius) [inline incubator/node/README.md]: It clones and deploy**s** a Node.js application from a git repository.

(I think we can omit the part about the revision)

File: incubator/node/README.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+## Installing the Chart
+
+To install the chart with the release name `my-release`:
+
+```console
+$ helm install --name my-release .
+```
+
+The command deploys node on the Kubernetes cluster in the default configuration. The [configuration](#configuration) section lists the parameters that can be configured during installation.
```
Comment (prydonius) [inline incubator/node/README.md]: capitalise node

File: incubator/node/templates/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        image: "{{ .Values.image }}"
+        imagePullPolicy: {{ default "" .Values.imagePullPolicy | quote }}
+        env:
+        - name: GIT_REPO
+          value: {{ .Values.repository }}
+        command:
+        - "/bin/bash"
+        - "-ec"
+        - |
+          DIR=$(echo $GIT_REPO | awk -F '/' '{print $NF }' | sed 's/.git$//g')
```
Comment (prydonius) [inline incubator/node/templates/deployment.yaml]: on second thought I think we should use init containers here, that'll let you clone the repo and set the right permissions and run npm install in the persisted volume. Also, the gitRepo volume isn't really supported and is just for examples.

File: incubator/node/values.yaml [hunk 9]
```diff
@@ -0,0 +1,59 @@
+## Bitnami node image version
+## ref: https://hub.docker.com/r/bitnami/node/tags/
+##
+image: bitnami/node:7.0.0-r1
+
+
```
Comment (prydonius) [inline incubator/node/values.yaml]: remove unnecessary whitespace

File: incubator/node/README.md [hunk 10]
```diff
@@ -12,7 +12,7 @@ $ helm install .
 
 This chart bootstraps a [Node](https://github.com/bitnami/bitnami-docker-node) deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.
 
-It clones and deploy a nodejs application  from a git repository and  defined revision.
+It clones and deploys a Node.js application  from a git repository.
```
Comment (prydonius) [inline incubator/node/README.md]: There's two spaces between `application` and `from` here

File: incubator/node/README.md [hunk 11]
```diff
@@ -65,11 +63,11 @@ Specify each parameter using the `--set key=value[,key=value]` argument to `helm
 
 ```console
 $ helm install --name my-release \
-  --set repository=git@github.com:jbianquetti-nami/simple-node-app.git,mariadb.mariadbRootPassword=secretpassword \
+  --set repository=http://github.com/jbianquetti-nami/simple-node-app.git,mariadb.mariadbRootPassword=secretpassword \
```
Comment (prydonius) [inline incubator/node/README.md]: https?

File: incubator/node/templates/deployment.yaml [hunk 12]
```diff
@@ -13,6 +13,20 @@ spec:
     metadata:
       labels:
         app: {{ template "fullname" . }}
+      annotations:
+        pod.beta.kubernetes.io/init-containers: '[
+            {
+                "name": "git-clone-app",
+                "image": "{{ .Values.image }}", 
+                "command": [ "/bin/sh", "-c" , "git clone {{ .Values.repository }} /app" ],
```
Comment (prydonius) [inline incubator/node/templates/deployment.yaml]: can we move the chown and `npm install` to here? So we can keep the main container only a single command?

File: incubator/node/templates/deployment.yaml [hunk 13]
```diff
@@ -25,9 +39,8 @@ spec:
         - "/bin/bash"
         - "-ec"
         - |
-          DIR=$(echo $GIT_REPO | awk -F '/' '{print $NF }' | sed 's/.git$//g')
+          git checkout {{ .Values.revision }}
           chown -R bitnami:daemon /app 
-          cd /app/$DIR
           su  bitnami -c 'PATH=$PATH:/opt/bitnami/node/bin/ npm install' 
           su  bitnami -c 'PATH=$PATH:/opt/bitnami/node/bin/ npm start'
```
Comment (prydonius) [inline incubator/node/templates/deployment.yaml]: can we use the runAsUser security context instead so we can still run this as PID 1?

File: incubator/node/templates/deployment.yaml [hunk 14]
```diff
@@ -52,6 +65,4 @@ spec:
           mountPath: /app
       volumes:
       - name: node-data
-        gitRepo:
-          repository: {{ .Values.repository }}
-          revision: {{ .Values.revision }}
+        emptyDir: {}
```
Comment (prydonius) [inline incubator/node/templates/deployment.yaml]: we should use the standard persistent settings here

File: incubator/node/README.md (hunk 10 above)
Comment (jbianquetti-nami) [inline incubator/node/README.md]: Done.

File: incubator/node/README.md [hunk 15]
```diff
@@ -27,7 +27,9 @@ To install the chart with the release name `my-release`:
 $ helm install --name my-release .
 ```
 
-The command deploys node on the Kubernetes cluster in the default configuration. The [configuration](#configuration) section lists the parameters that can be configured during installation.
+The command deploys Node.js on the Kubernetes cluster in the default configuration. The [configuration](#configuration) section lists the parameters that can be configured during installation. Also includes support for MariaDB chart out of the box. 
+
+Due that the Helm Chart clones the application on the /app volume while the container is initializing, there's not a hard dependency on Persistent Volumes.
```
Comment (jbianquetti-nami) [inline incubator/node/README.md]: Done.

File: incubator/node/README.md (hunk 11 above)
Comment (jbianquetti-nami) [inline incubator/node/README.md]: Done.

File: incubator/node/templates/deployment.yaml (hunk 13 above)
Comment (jbianquetti-nami) [inline incubator/node/templates/deployment.yaml]: Not possible unless we move the steps to the init-container. See previous  comment.

File: incubator/node/templates/deployment.yaml (hunk 14 above)
Comment (jbianquetti-nami) [inline incubator/node/templates/deployment.yaml]: Done.

File: incubator/node/Chart.yaml [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+- nodejs
+- git
+home: http://nodejs.org/
+sources:
+- https://github.com/bitnami/bitnami-docker-node
+maintainers:
+- name: Bitnami
+  email: containers@bitnami.com
+engine: gotpl
+icon: https://d33np9n32j53g7.cloudfront.net/assets/stacks/nodejs/img/nodejs-stack-110x117-331955cd5e874f1ebd26ff9bb0c5af4c.png
```
Comment (prydonius) [inline incubator/node/Chart.yaml]: This should be `https://bitnami.com/assets/stacks/nodejs/img/nodejs-stack-110x117.png`. In the future, just convert the cloudfront url to bitnami.com, and remove the checksum at the end.

File: incubator/node/templates/NOTES.txt [hunk 17]
```diff
@@ -0,0 +1,22 @@
+
+1. Get the node URL by running:
```
Comment (prydonius) [inline incubator/node/templates/NOTES.txt]: > Get the URL of your Node app?

File: incubator/node/templates/deployment.yaml (hunk 12 above)
Comment (prydonius) [inline incubator/node/templates/deployment.yaml]: What was the issue with stability?
